C/jets-secp256k1.c

```diff
@@ -0,0 +1,218 @@
+ *
+ * Precondition: '*dst' is a valid write frame for 960 more cells;
+ *               NULL != r;
+ */
+static inline void write_gej(frameItem* dst, const secp256k1_gej* r) {
+  write_fe(dst, &r->x);
+  write_fe(dst, &r->y);
+  if (r->infinity) {
+    write32s(dst, (uint32_t[10]){0}, 10);
+  } else {
+    write_fe(dst, &r->z);
+  }
+}
+
+/* Read a secp256k1 scalar element value from the 'src' frame, advancing the cursor 256 cells.
+ * The secp256k1 8x32 scalar representation puts the 32 least significant bytes into the first array element;
+ * However Simplicity uses a standard big-endian 256-bit word to represent scalar values.
+ * Thus it is necessary to fill the secp256k1 scalar array in reverse order.
+ *
+ * Precondition: '*src' is a valid read frame for 256 more cells;
+ *               NULL != r;
+ */
+static inline void read_scalar(secp256k1_scalar* r, frameItem* src) {
+  r->d[7] = (uint32_t)read32(src);
+  r->d[6] = (uint32_t)read32(src);
+  r->d[5] = (uint32_t)read32(src);
+  r->d[4] = (uint32_t)read32(src);
+  r->d[3] = (uint32_t)read32(src);
+  r->d[2] = (uint32_t)read32(src);
+  r->d[1] = (uint32_t)read32(src);
+  r->d[0] = (uint32_t)read32(src);
+}
+
+/* Write a secp256k1 scalar element value to the 'dst' frame, advancing the cursor 256 cells.
+ * The secp256k1 8x32 scalar representation puts the 32 least significant bytes into the first array element;
+ * However Simplicity uses a standard big-endian 256-bit word to represent scalar values.
+ * Thus it is necessary to read from the secp256k1 scalar array in reverse order.
+ *
+ * Precondition: '*dst' is a valid write frame for 256 more cells;
+ *               NULL != r;
+ */
+static inline void write_scalar(frameItem* dst, const secp256k1_scalar* r) {
+  write32(dst, r->d[7]);
+  write32(dst, r->d[6]);
+  write32(dst, r->d[5]);
+  write32(dst, r->d[4]);
+  write32(dst, r->d[3]);
+  write32(dst, r->d[2]);
+  write32(dst, r->d[1]);
+  write32(dst, r->d[0]);
+}
+
+bool fe_sqrt(frameItem* dst, frameItem src, const txEnv* env) {
+  (void) env; // env is unused;
+
+  secp256k1_fe r, a;
+  read_fe(&a, &src);
+  int result = secp256k1_fe_sqrt_var(&r, &a);
+  if (writeBit(dst, result)) {
+    write_fe(dst, &r);
+  } else {
+    skip_fe(dst);
+  }
+  return true;
+}
+
+bool offsetPoint(frameItem* dst, frameItem src, const txEnv* env) {
+  (void) env; // env is unused;
+
+  secp256k1_gej r, a;
+  secp256k1_ge b;
+  secp256k1_fe rzr;
+  read_gej(&a, &src);
+  read_ge(&b, &src);
+  secp256k1_gej_add_ge_var(&r, &a, &b, &rzr);
+  write_fe(dst, &rzr);
+  write_gej(dst, &r);
+  return true;
+}
+
+static struct {
+   secp256k1_ecmult_context ctx;
+   char alloc[SECP256K1_ECMULT_CONTEXT_PREALLOCATED_SIZE];
+   bool initialized;
+} ecmult_static;
+
+/* This function will initialize the 'ecmult_static' global variable if it hasn't already been initialized. */
+static void ensure_ecmult_static(void) {
```

Comment:
The thread-unsafety of this really makes me uncomfortable, though I can't think of any other way to do this. Maybe putting a secp context into `txEnv`?

Reply:
This is a more general problem with the entire library.  I've created [a separate issue](https://github.com/ElementsProject/simplicity/issues/33) to track it .